torch/csrc/jit/interpreter.cpp

```diff
@@ -213,46 +485,83 @@ Operation getOperation(jit::Node *node) {
+        stack.push_back(t);
+        return 0;
+      };
+    } else {
+      auto t = value->t(kvalue);
+      return [t](Stack & stack) {
+        stack.push_back(t);
+        return 0;
+      };
+    }
   IR_ELSEIF(Undefined)
-    return [](const list_of_retainable & inputs, list_of_retainable & outputs) {
-      outputs.push_back(toRetainableSteal(at::Tensor()));
+    return [](Stack & stack) {
+      stack.push_back(at::Tensor());
+      return 0;
     };
   IR_ELSEIF(ReplaceIfUndef)
-    return [](const list_of_retainable & inputs, list_of_retainable & outputs) {
-      auto result = inputs[0];
-      //TODO: refcounting stuff here is ugly but TensorTemporary is not
-      //present. Consider whether we
-      // 1. expose tensor temporary here
-      // 2. keep as is
-      // 3. remove all of this retainable stuff anyway since the new
-      // execution paths do not need handle types.
-      // Note that list_of_retainable is painful because it is yet another
-      // list of pointers that require needless copies.
-      if(result == at::UndefinedTensor::singleton()) {
-        result = inputs[1];
+    return [](Stack & stack) {
+      auto alternate = pop(stack);
+      auto result = pop(stack);
+      if(result.defined()) {
+        stack.push_back(std::move(result));
+      } else {
+        stack.push_back(std::move(alternate));
       }
-      result->retain();
-      outputs.push_back(result);
+      return 0;
     };
   IR_ELSEIF(GraphExecutor)
     GraphExecutor executor(value->g(kSubgraph));
-    return [=](const list_of_retainable & inputs, list_of_retainable & outputs) mutable {
+    auto num_inputs = value->inputs().size();
+    return [=](Stack& stack) mutable {
       autograd::profiler::RecordFunction record("GraphExecutor");
-      tensor_list tinputs = unsafeToTensorListShare(inputs);
+      auto inputs = last(stack, num_inputs);
+      variable_tensor_list tinputs(Stack(inputs.begin(), inputs.end()));
+      drop(stack, num_inputs);
+      //TODO: has graph executor work from a stack as well
       variable_tensor_list toutputs = executor.run(variable_tensor_list(std::move(tinputs)));
-      moveToListOfRetainables(std::move(toutputs), outputs);
+      stack.insert(stack.end(), toutputs.begin(), toutputs.end());
+      return 0;
+    };
+
+
+  // Load x, y
+  // loads values from registers onto the stack, the actual callback does
+  // nothing since the stack manipulation is already encoded in inst.inputs
+  // and inst.outputs
+  IR_ELSEIF(Load)
+    return [=](Stack& stack) {
+      return 0;
+    };
+
+  // x, y = Store
+  // stores values from stack into registers, the actual callback does
+  // nothing since the stack manipulation is already encoded in inst.inputs
+  // and inst.outputs
+  IR_ELSEIF(Store)
+    return [=](Stack& stack) {
+      return 0;
+    };
+  IR_ELSEIF(Drop)
+    auto N = value->inputs().size();
+    return [=](Stack& stack) {
+      drop(stack, N);
+      return 0;
```

Comment:
So the only reason we're inserting `Drop` nodes is to have every value "used" at least once, such that `findLastUses` will free it? I only realized it at this point, it would be good to mention that somewhere earlier. From what I've seen previously I've been assuming it will somehow reach into the register list and remove stuff, while it really is a dummy operation with no outputs

Reply:
Yes, I've clarified their use in the comment describing preprocessing.